client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -0,0 +1,236 @@
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+        this.tokenRefreshThreshold = tokenRefreshThreshold;
+    }
+
+    /**
+     * Returns the delegation token. It returns existing delegation token if it is not close to expiry. If the token
+     * is close to expiry, it obtains a new delegation token and returns that one instead.
+     *
+     * @return String the delegation token JWT compact value
+     */
+    @Override
+    public String retrieveToken() {
+        if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry");
+            return refreshToken();
+        } else {
+            return delegationToken.get().getValue();
+        }
+    }
+
+    @Override
+    public String refreshToken() {
+        log.trace("Entered refreshToken().");
+        DelegationToken existingToken = this.getCurrentToken();
+        if (isTokenBeingRefreshed.get()) {
+            log.debug("Token is already under refresh.");
+
+            // Wait for the token to finish refreshing
+            while (!isTokenBeingRefreshed.get()) {
+                try {
+                    Thread.sleep(200);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        } else {
+            if (isTokenBeingRefreshed.compareAndSet(false, true)) {
+                log.debug("Resetting token");
+                resetToken(existingToken, newToken());
+            } else {
+                log.debug("Token refresh status setting to 'true' failed.");
+            }
+        }
+        String result = delegationToken.get().getValue();
+        log.trace("Exiting refreshToken()");
+        return result;
+    }
+
+    protected DelegationToken newToken() {
+        String token = Futures.getAndHandleExceptions(
+                controllerClient.getOrRefreshDelegationTokenFor(scopeName, streamName), RuntimeException::new);
+        return new DelegationToken(token, extractExpirationTime(token));
+    }
+
+    protected DelegationToken getCurrentToken() {
+        return this.delegationToken.get();
+    }
+
+    private boolean isTokenNearingExpiry() {
+        Long currentTokenExpirationTime = this.delegationToken.get().getExpiryTime();
+
+        // currentTokenExpirationTime can be null if the server returns a null delegation token
+        return currentTokenExpirationTime != null && isWithinRefreshThreshold(currentTokenExpirationTime);
+    }
+
+    private boolean isWithinRefreshThreshold(Long expirationTime) {
+        Preconditions.checkNotNull(expirationTime);
```

Comment:
for private method, it's best to use `assert expirationTime != null`. This will be checked in unit and system tests, but it won't be enabled in production so it will make your code faster at runtime.

Reply:
Yes, that makes sense to use asserts for private methids. I've removed the preconditions check and replaced it with an `assert`, in the latest commit.